tools/publisher/src/cargo.rs

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Module for interacting with Cargo.
+
+use anyhow::{Context, Result};
+use std::path::{Path, PathBuf};
+use std::process::{Command, Output};
+
+macro_rules! cmd {
+    [ $( $x:expr ),* ] => {
+        {
+            let mut cmd = Cmd::new();
+            $(cmd.push($x);)*
+            cmd
+        }
+    };
+}
+
+/// Confirms that cargo exists on the path.
+pub async fn confirm_installed_on_path() -> Result<()> {
+    cmd!["cargo", "--version"]
+        .spawn()
+        .await
+        .context("cargo is not installed on the PATH")?;
+    Ok(())
+}
+
+/// Returns a `Cmd` that, when spawned, will asynchronously run `cargo publish` in the given crate path.
+pub fn publish_task(crate_path: &Path) -> Cmd {
+    cmd!["cargo", "publish"].working_dir(crate_path)
+}
+
+#[derive(Default)]
+pub struct Cmd {
+    parts: Vec<String>,
+    working_dir: Option<PathBuf>,
+}
+
+impl Cmd {
+    fn new() -> Cmd {
+        Default::default()
+    }
+
+    fn push(&mut self, part: impl Into<String>) {
+        self.parts.push(part.into());
+    }
+
+    fn working_dir(mut self, working_dir: impl AsRef<Path>) -> Self {
+        self.working_dir = Some(working_dir.as_ref().into());
+        self
+    }
+
+    /// Returns a plan string that can be output to the user to describe the command.
+    pub fn plan(&self) -> String {
+        let mut plan = String::new();
+        if let Some(working_dir) = &self.working_dir {
+            plan.push_str(&format!("[in {:?}]: ", working_dir));
+        }
+        plan.push_str(&self.parts.join(" ").to_string());
+        plan
+    }
+
+    /// Runs the command asynchronously.
+    pub async fn spawn(mut self) -> Result<Output> {
```

Comment:
any reason to do this async? Seems like we may want to block on these commands running